hudhook/src/hooks/opengl3.rs

```diff
@@ -0,0 +1,285 @@
+            trace!("GetWindowRect error: {:x}", GetLastError().0);
+        }
+
+        // Update the delta time of ImGui as to tell it how long has elapsed since the
+        // last frame
+        let last_frame = LAST_FRAME.get_or_insert_with(|| Mutex::new(Instant::now())).get_mut();
+        let now = Instant::now();
+        self.ctx.io_mut().update_delta_time(now.duration_since(*last_frame));
+        *last_frame = now;
+
+        let mut ui = self.ctx.frame();
+
+        IMGUI_RENDER_LOOP.get_mut().unwrap().render(&mut ui, &self.flags);
+        self.renderer.render(ui);
+    }
+
+    unsafe fn cleanup(&mut self) {}
+}
+
+impl ImguiWindowsEventHandler for ImguiRenderer {
+    fn io(&self) -> &imgui::Io {
+        self.ctx.io()
+    }
+
+    fn io_mut(&mut self) -> &mut imgui::Io {
+        self.ctx.io_mut()
+    }
+
+    fn focus(&self) -> bool {
+        self.flags.focused
+    }
+
+    fn focus_mut(&mut self) -> &mut bool {
+        &mut self.flags.focused
+    }
+
+    fn wnd_proc(&self) -> WndProcType {
+        self.wnd_proc
+    }
+}
+unsafe impl Send for ImguiRenderer {}
+unsafe impl Sync for ImguiRenderer {}
+
+/// Stores hook detours and implements the [`Hooks`] trait.
+pub struct OpenGL3Hooks {
+    #[allow(dead_code)]
+    hook_opengl_wgl_swap_buffers: RawDetour,
+}
+
+impl OpenGL3Hooks {
+    /// # Safety
+    ///
+    /// Is most likely undefined behavior, as it modifies function pointers at
+    /// runtime.
+    pub unsafe fn new<T: 'static>(t: T) -> Self
+    where
+        T: ImguiRenderLoop + Send + Sync,
+    {
+        // Grab the addresses
+        let hook_opengl_swapbuffers_address = get_opengl_wglswapbuffers_addr();
+
+        // Create detours
+        let hook_opengl_wgl_swap_buffers = RawDetour::new(
+            hook_opengl_swapbuffers_address as *const _,
+            imgui_opengl3_wglSwapBuffers_impl as *const _,
+        )
+        .expect("opengl32.wglSwapBuffers hook");
+
+        // Initialize the render loop and store detours
+        IMGUI_RENDER_LOOP.get_or_init(|| Box::new(t));
+        TRAMPOLINE.get_or_init(|| std::mem::transmute(hook_opengl_wgl_swap_buffers.trampoline()));
+
+        Self { hook_opengl_wgl_swap_buffers }
+    }
+}
+
+impl Hooks for OpenGL3Hooks {
+    unsafe fn hook(&self) {
+        for hook in [&self.hook_opengl_wgl_swap_buffers] {
+            if let Err(e) = hook.enable() {
+                error!("Couldn't enable hook: {e}");
+            }
+        }
+    }
+
+    unsafe fn unhook(&mut self) {
+        for hook in [&self.hook_opengl_wgl_swap_buffers] {
+            if let Err(e) = hook.disable() {
+                error!("Couldn't disable hook: {e}");
+            }
+        }
+
+        if let Some(renderer) = IMGUI_RENDERER.take() {
+            renderer.lock().cleanup();
+        }
+        drop(IMGUI_RENDER_LOOP.take());
+    }
+}
+
+// Get the address of wglSwapBuffers in opengl32.dll
+unsafe fn get_opengl_wglswapbuffers_addr() -> OpenGl32wglSwapBuffers {
```

Comment:
Can we move this before the `OpenGL3Hooks` struct definition? I'd prefer having dependent functions below dependencies.

Reply:
Done